src/sources/pulsar.rs

```diff
@@ -0,0 +1,451 @@
+                                        if let Event::Log(ref mut log) = event {
+                                            log.try_insert(
+                                                crate::config::log_schema().source_type_key(),
+                                                bytes::Bytes::from("pulsar"),
+                                            );
+                                            log.try_insert(crate::config::log_schema().timestamp_key(), now);
+                                        }
+                                        event
+                                    });
+
+                                    for event in events {
+                                        yield event;
+                                    }
+                                }
+                                Err(error) => {
+                                    // Error is logged by `crate::codecs`, no further
+                                    // handling is needed here.
+                                    if !error.can_continue() {
+                                        break;
+                                    }
+                                }
+                            }
+                        }
+                        }.boxed();
+
+                        finalize_event_stream(consumer.clone(), &finalizer, &mut out, stream, msg.clone()).await;
+                    }
+                    Err(error) => {
+                        emit!(PulsarReadError { error })
+                    }
+                }
+            },
+        }
+    }
+
+    Ok(())
+}
+
+/// Send the event stream created by the framed read to the `out` stream.
+async fn finalize_event_stream(
+    consumer: Arc<Mutex<Consumer<String, TokioExecutor>>>,
+    finalizer: &Option<OrderedFinalizer<FinalizerEntry>>,
+    out: &mut SourceSender,
+    mut stream: std::pin::Pin<Box<dyn futures_util::Stream<Item = Event> + Send + '_>>,
+    message: pulsar::consumer::Message<std::string::String>,
+) {
+    match finalizer {
+        Some(finalizer) => {
+            let (batch, receiver) = BatchNotifier::new_with_receiver();
+            let mut stream = stream.map(|event| event.with_batch_notifier(&batch));
+
+            match out.send_event_stream(&mut stream).await {
+                Err(error) => {
+                    emit!(StreamClosedError { error, count: 1 });
+                }
+                Ok(_) => {
+                    finalizer.add(FinalizerEntry{ consumer, message }, receiver);
+                }
+            }
+        }
+        None => match out.send_event_stream(&mut stream).await {
+            Err(error) => {
+                emit!(StreamClosedError { error, count: 1 });
+            }
+            Ok(_) => {
+                if let Err(error) = consumer.lock().await.ack(&message).await {
+                    emit!(PulsarAcknowledgmentError { error });
+                }
+            }
+        },
+    }
```

Comment:
This could have the duplication reduced with `map` adapters, not sure if it's simpler:
```suggestion
    if let Err(error) = match finalizer {
        Some(finalizer) => {
            let (batch, receiver) = BatchNotifier::new_with_receiver();
            let mut stream = stream.map(|event| event.with_batch_notifier(&batch));

            out.send_event_stream(&mut stream).await.map(|_|
                finalizer.add(FinalizerEntry{ consumer, message }, receiver)
            )
        }
        None => out.send_event_stream(&mut stream).await.map(|_| {
            if let Err(error) = consumer.lock().await.ack(&message).await {
                emit!(PulsarAcknowledgmentError { error });
            }
        }),
    } {
        emit!(StreamClosedError { error, count: 1 });
    }
```